### PR TITLE
PS-137 Doorman update

### PR DESCRIPTION
**Pull request by pavel-farkas**
Updates:

- caller greeting is now configurable
- b-leg is always originated by the same FS node which processes a-leg, this allows to store recorded caller name only locally in /tmp folder and recorded name is always available to b-leg
- doorman menu has been split: callee greeting and caller name playback create one part, which is not interruptible by DTMF, doorman options create second part, where DTMF collection takes place and only this part is repeated when callee selects wrong option
- removed code for cleaning of recorded names, it won't work as intended, cf_doorman would not have right access rights to delete files created by Freeswitch. Instead, /tmp folder of FS nodes is cleaned regularly by cronjob running `tmpwatch`
- caller prompts are now taken from kazoo-sounds, tts is used only for caller greeting when it is configured in doorman document - this allows to override default prerecorded greeting
- callee option "connect and record" has been removed per client request 
- added failover strategy config parameter: what to do if the callee doesn't answer or is busy, default action is **hangup**, alternative is go to **voicemail**
- if no vmbox is configured, doorman will try to find vmbox associated with configured user or owner of configured device
- added config parameter **max_menu_attempts**: how many times to allow the callee to choose from the menu

### CODE REVIEW

File: src/module/cf_doorman.erl [hunk 1]
```diff
@@ -350,30 +352,19 @@ validate_originate_resp(_Event, Resp, _CalleeCall, _DoormanProc) ->
...
             {'error', 'callee_hungup'};
-        {'ok', ?DTMF_CONNECT_CALL=Choice} -> {'ok', Choice, CalleeCall};
-        {'ok', ?DTMF_CONNECT_RECORD=Choice} -> {'ok', Choice, CalleeCall};
-        {'ok', ?DTMF_VM=Choice} -> {'ok', Choice, CalleeCall};
-        {'ok', ?DTMF_HANGUP=Choice} ->
-            lager:info("callee has selected not to take the call (pressed ~p)", [Choice]),
+        {'ok', Option} when Option =:= ?DTMF_CONNECT_CALL
+                            orelse Option =:= ?DTMF_CONNECT_RECORD
+                            orelse Option =:= ?DTMF_VM -> {'ok', Option, CalleeCall};
+        {'ok', ?DTMF_HANGUP} ->
```
Comment (jamesaimonetti) [inline src/module/cf_doorman.erl:399]: If you re-order this clause above the others, you don't need the orelse block anymore, right?

Comment (pavel-farkas) [reply src/module/cf_doorman.erl:399]: Correct, fixed, thank you.

File: src/module/cf_doorman.erl [hunk 2]
```diff
@@ -442,6 +442,7 @@ build_originate_request(Data, CalleeCall, Endpoints, _CallerCallId) ->
                                           ]}
       ,{<<"Ignore-Early-Media">>, 'true'}
       ,{<<"Originate-Immediate">>, 'false'}
+      ,{<<"Switch-Nodename">>, kapps_call:switch_nodename(CallerCall)}
```
Comment (jamesaimonetti) [inline src/module/cf_doorman.erl:527]: As far as I recall, this key does not put the origination leg on the same switch as the caller. Why not use `Existing-Call-ID`?

See https://github.com/2600hz/kazoo-core/blob/master/kazoo_amqp/src/api/kapi_resource.erl#L143

**Review by jamesaimonetti — CHANGES_REQUESTED** [review]

File: src/module/cf_doorman.erl (hunk 2 above)
Comment (pavel-farkas) [reply src/module/cf_doorman.erl:527]: Hi James,
yes, I started with `Existing-Call-ID`, but then I realised that everything what is played to callee as 'doorman menu' -prompts, recorded name, options - is also played to caller. Even if those two call legs were not bridged yet. Only solution how to fix it was not using `Existing-Call-ID`. 

Prompts and record playback have parameter to choose call leg for playback, but it was working only for value `both`. As far as I know it should accept also `aleg` and `bleg` values, but result for these two was always silence on both ends.

So I started to look for alternative. I've found your code in `cb_quickcall` module where you set both `Existing-Call-ID` and `Switch-Nodename` together:
https://github.com/2600hz/kazoo-crossbar/blob/master/src/modules/cb_quickcall.erl#L349

But I was mainly inspired by this code:
https://github.com/2600hz/kazoo-core/blob/master/kazoo_amqp/src/api/kapi_resource.erl#L395
where `Switch-Nodename` value is used to create routing key for originate amqp request. Isn't that enough to make sure that request will be processed by required FS node? If not, I will return to `Existing-Call-ID` and try to find another workaround.

Nevertheless, with `Switch-Nodename` set, both call legs were always processed by the same FS node during my tests, so I was convinced that I'm using right parameter.

**Comment by pavel-farkas** [comment]
Hi @jamesaimonetti,
I'd like to ask for review of updated Doorman code. It now contains all changes that customer required before accepting the final solution. They are using it now for almost two months in production, no complains so far, so I believe it is stable enough. 
Thank you.

**Comment by jamesaimonetti** [comment]
Yes, thanks for the reminder @pavel-farkas

**Pull request merged**